optaplanner-core/src/main/java/org/optaplanner/core/api/score/holder/AbstractScoreHolder.java

```diff
@@ -67,16 +68,44 @@ public boolean isConstraintMatchEnabled() {
     // Worker methods
     // ************************************************************************
 
+    protected class MultiLevelConstraintUndoListener implements ActivationUnMatchListener {
+
+        private Map<Integer, ActivationUnMatchListener> activationUnMatchListenerMap;
+
+        public MultiLevelConstraintUndoListener(int scoreLevel, ActivationUnMatchListener activationUnMatchListener) {
+            this.activationUnMatchListenerMap = new HashMap<Integer, ActivationUnMatchListener>();
```

Comment:
D: if there's only 1 undo listener, put it in directly, only use the MultiLevelConstraintUndoListener if there are 2 score levels. But that won't work as we'll need to know the undo listener's score level

A: +1 Good idea. HashMap's automatically expand anyway 
B: -1 If we have a bendable with 1000 levels, then we have a serious scalability problem: even with 10k matches each using just 1 level, we reserve 10m memory. Also, the AbstractScoreHolder needs to add an abstract method to get the score level size from it's children.
C: +0
D: +0 painful